// mw-style-react/build/Table/Table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class, _class2, _temp;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _Table = require('./Table.style');

var _Table2 = _interopRequireDefault(_Table);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Table = (_dec = (0, _reactJss2.default)(_Table2.default), _dec(_class = (_temp = _class2 = function (_PureComponent) {
  (0, _inherits3.default)(Table, _PureComponent);

  function Table(props) {
    (0, _classCallCheck3.default)(this, Table);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Table.__proto__ || (0, _getPrototypeOf2.default)(Table)).call(this, props));

    _this.state = {
      columnSize: {}
    };
    _this.handleColumnSize = _this.handleColumnSize.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Table, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var _props = this.props,
          scrolled = _props.scrolled,
          visibility = _props.visibility,
          paddingSize = _props.paddingSize;

      return { scrolled: scrolled, visibility: visibility, paddingSize: paddingSize };
    }
  }, {
    key: 'handleColumnSize',
    value: function handleColumnSize(columnSize) {
      this.setState({
        columnSize: columnSize
      });
    }
  }, {
    key: 'renderTable',
    value: function renderTable(tableClasses, children) {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        { className: tableClasses },
        _react2.default.Children.map(children, function (child) {
          return _react2.default.cloneElement(child, {
            getColumnSize: _this2.handleColumnSize,
            columnSize: _this2.state.columnSize
          });
        })
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          className = _props2.className,
          classes = _props2.classes,
          visibility = _props2.visibility,
          children = _props2.children,
          scrolled = _props2.scrolled;


      var tableClasses = (0, _clsx2.default)(classes.table, (0, _defineProperty3.default)({}, '' + className, !scrolled), { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' });

      var tableWrapClasses = (0, _clsx2.default)(classes.tableWrap, className, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' }, { scrolled: scrolled });

      return scrolled ? _react2.default.createElement(
        'div',
        { className: tableWrapClasses },
        this.renderTable(tableClasses, children)
      ) : this.renderTable(tableClasses, children);
    }
  }]);
  return Table;
}(_react.PureComponent), _class2.childContextTypes = {
  scrolled: _propTypes2.default.bool,
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  paddingSize: _propTypes2.default.oneOf(['small', 'medium'])
}, _temp)) || _class);


Table.defaultProps = {
  visibility: 'visible',
  scrolled: false,
  paddingSize: 'medium'
};

Table.propTypes = {
  /** React-JSS injects a classes, which is a simple map of rule names and generated class names */
  classes: _propTypes2.default.object,
  /** additional class name */
  className: _propTypes2.default.string,
  /** table's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** table's horizontal padding */
  paddingSize: _propTypes2.default.oneOf(['small', 'medium']),
  /** table's Content */
  children: _propTypes2.default.node.isRequired,
  /** is scrollable */
  scrolled: _propTypes2.default.bool
};

exports.default = Table;
