// mw-style-react/build/Tab/Tab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _Tab = require('./Tab.style');

var _Tab2 = _interopRequireDefault(_Tab);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Tab = (_dec = (0, _reactJss2.default)(_Tab2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Tab, _PureComponent);

  function Tab(props) {
    (0, _classCallCheck3.default)(this, Tab);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Tab.__proto__ || (0, _getPrototypeOf2.default)(Tab)).call(this, props));

    _this.handleChange = _this.handleChange.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Tab, [{
    key: 'handleChange',
    value: function handleChange(value) {
      var _props = this.props,
          id = _props.id,
          onChange = _props.onChange;

      var error = false;
      onChange({ id: id, value: value, error: error });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          children = _props2.children,
          className = _props2.className,
          id = _props2.id,
          classes = _props2.classes,
          size = _props2.size,
          underline = _props2.underline,
          error = _props2.error,
          value = _props2.value,
          type = _props2.type,
          tabItemIndent = _props2.tabItemIndent;


      var tabClasses = (0, _clsx2.default)(className, classes.tabContainer, classes[size], classes[type], {
        bottomLine: !!underline
      });
      var childrenTabs = Array.isArray(children) ? children : [children];
      var visibilityTabs = childrenTabs.filter(function (c) {
        return c.props.visibility !== 'hidden';
      });

      return _react2.default.createElement(
        'div',
        { id: id, className: tabClasses },
        _react2.default.createElement(
          'div',
          { ref: function ref(container) {
              return _this2.container = container;
            }, className: (0, _clsx2.default)(classes.tabs) },
          _react2.default.Children.map(visibilityTabs, function (tab) {
            return _react2.default.cloneElement(tab, {
              itemRef: function itemRef(el) {
                _this2['itemRef' + tab.props.value] = el;
              },
              onChange: _this2.handleChange,
              activeItem: value,
              error: error,
              size: size,
              type: type,
              tabItemIndent: tabItemIndent
            });
          })
        )
      );
    }
  }]);
  return Tab;
}(_react.PureComponent)) || _class);


Tab.defaultProps = {
  id: null,
  children: null,
  value: null,
  visibility: 'visible',
  type: 'fixed',
  size: 'medium',
  tabItemIndent: 'medium',
  underline: false,
  onChange: function onChange() {}
};

Tab.propTypes = {
  classes: _propTypes2.default.object,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** Tab's id */
  id: _propTypes2.default.string,
  /** Tab's bottom line */
  type: _propTypes2.default.oneOf(['fixed', 'auto']),
  /** Tab's bottom line */
  underline: _propTypes2.default.bool,
  /** Tab options */
  children: _propTypes2.default.node,
  /** Tab's active tab id */
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /** Tab's callback */
  onChange: _propTypes2.default.func,
  /** Tab's size */
  size: _propTypes2.default.oneOf(['small', 'medium']),
  /** TabItem indent */
  tabItemIndent: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  /** Error Tab id */
  error: _propTypes2.default.bool
};

exports.default = Tab;
