// mw-style-react/build/Autocomplete/AutocompleteItem/AutocompleteItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _AutocompleteItem = require('./AutocompleteItem.style');

var _AutocompleteItem2 = _interopRequireDefault(_AutocompleteItem);

var _Label = require('../../Label');

var _Label2 = _interopRequireDefault(_Label);

var _Avatar = require('../../Avatar');

var _Avatar2 = _interopRequireDefault(_Avatar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AutocompleteItem = (_dec = (0, _reactJss2.default)(_AutocompleteItem2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(AutocompleteItem, _PureComponent);

  function AutocompleteItem(props) {
    (0, _classCallCheck3.default)(this, AutocompleteItem);

    var _this = (0, _possibleConstructorReturn3.default)(this, (AutocompleteItem.__proto__ || (0, _getPrototypeOf2.default)(AutocompleteItem)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(AutocompleteItem, [{
    key: 'handleClick',
    value: function handleClick() {
      var _props = this.props,
          id = _props.id,
          value = _props.value,
          avatar = _props.avatar,
          visibility = _props.visibility;

      if (visibility === 'disabled') return;
      this.props.onClick({ id: id, value: value, avatar: avatar, error: false });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          id = _props2.id,
          classes = _props2.classes,
          className = _props2.className,
          visibility = _props2.visibility,
          value = _props2.value,
          avatar = _props2.avatar,
          size = _props2.size,
          activeItem = _props2.activeItem;


      return _react2.default.createElement(
        'div',
        {
          name: id,
          className: (0, _clsx2.default)(className, size, classes.autocompleteItem, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' }, { active: activeItem === id }, { unActive: activeItem && activeItem !== id }),
          onClick: this.handleClick
        },
        _react2.default.createElement(
          'div',
          { className: 'nameContainer' },
          _react2.default.createElement(_Avatar2.default, { label: value, src: avatar, size: size }),
          _react2.default.createElement(_Label2.default, { className: 'label', value: value, color: 'inherit' })
        )
      );
    }
  }]);
  return AutocompleteItem;
}(_react.PureComponent)) || _class);


AutocompleteItem.defaultProps = {
  id: null,
  value: null,
  visibility: 'visible',
  leftIcon: null,
  rightIcon: null,
  error: false,
  avatar: '',
  activeItem: null,
  size: 'large',
  onClick: function onClick() {}
};

AutocompleteItem.propTypes = {
  classes: _propTypes2.default.object,
  /** autocomplete item's id */
  id: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /** autocomplete item's value */
  value: _propTypes2.default.string.isRequired,
  /** autocomplete item's avatar */
  avatar: _propTypes2.default.string,
  /** autocomplete item's size */
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  /** autocomplete item's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']).isRequired,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** onClick callback */
  onClick: _propTypes2.default.func,
  /** Active item id */
  activeItem: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number])
};

exports.default = AutocompleteItem;
