// mw-style-react/src/components/Badge/Badge.js
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import injectSheet from 'react-jss';
import cn from 'clsx';
import styles from './Badge.style';

@injectSheet(styles)
class Badge extends PureComponent {
  render() {
    const {
      classes,
      id,
      className,
      visibility,
      value,
      bgColor,
      size,
      onClick,
      borderColor,
    } = this.props;
    return (
      <div
        id={id}
        className={cn(
          className,
          classes.badge,
          classes[size],
          {hidden: visibility === 'hidden'},
          {disabled: visibility === 'disabled'},
          'badge',
        )}
        onClick={onClick}
        style={{
          borderColor,
          backgroundColor: bgColor,
        }}
      >
        {value}
      </div>
    );
  }
}

Badge.defaultProps = {
  visibility: 'visible',
  error: false,
  size: 'medium',
  onClick: () => {
  },
};

Badge.propTypes = {
  classes: PropTypes.object,
  /** additional external class name */
  className: PropTypes.string,
  /** Badge's visibility */
  visibility: PropTypes.oneOf(['visible', 'disabled', 'hidden']),
  /** Badge size */
  size: PropTypes.oneOf(['small', 'medium', 'large', 'xlarge', 'xxlarge']),
  /** Badge's id */
  id: PropTypes.string,
  /** Badge's value */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Badge border color */
  borderColor: PropTypes.string,
  /** Badge background color */
  bgColor: PropTypes.string,
  /** callback onClick */
  onClick: PropTypes.func,
};

export default Badge;
