// mw-style-react/build/Modal/ModalButtons/ModalButtons.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _ModalButtons = require('./ModalButtons.style');

var _ModalButtons2 = _interopRequireDefault(_ModalButtons);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ModalButtons = (_dec = (0, _reactJss2.default)(_ModalButtons2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(ModalButtons, _PureComponent);

  function ModalButtons() {
    (0, _classCallCheck3.default)(this, ModalButtons);
    return (0, _possibleConstructorReturn3.default)(this, (ModalButtons.__proto__ || (0, _getPrototypeOf2.default)(ModalButtons)).apply(this, arguments));
  }

  (0, _createClass3.default)(ModalButtons, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classes = _props.classes,
          id = _props.id,
          children = _props.children,
          visibility = _props.visibility,
          className = _props.className;


      return _react2.default.createElement(
        'div',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.buttonsWrapper, { disabled: visibility === 'disabled' }, { hidden: visibility === 'hidden' })
        },
        children
      );
    }
  }]);
  return ModalButtons;
}(_react.PureComponent)) || _class);


ModalButtons.defaultProps = {
  visibility: 'visible'
};

ModalButtons.propTypes = {
  classes: _propTypes2.default.object,
  /** ModalButtons's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** ModalButtons's id */
  id: _propTypes2.default.string,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** Buttons */
  children: _propTypes2.default.node.isRequired
};

exports.default = ModalButtons;
