// mw-style-react/src/components/Modal/ModalButtons/ModalButtons.js
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import injectSheet from 'react-jss';
import cn from 'clsx';
import styles from './ModalButtons.style';

@injectSheet(styles)
class ModalButtons extends PureComponent {
  render() {
    const {classes, id, children, visibility, className} = this.props;

    return (
      <div
        id={id}
        className={cn(
          className,
          classes.buttonsWrapper,
          {disabled: visibility === 'disabled'},
          {hidden: visibility === 'hidden'},
        )}
      >
        {children}
      </div>
    );
  }
}

ModalButtons.defaultProps = {
  visibility: 'visible',
};

ModalButtons.propTypes = {
  classes: PropTypes.object,
  /** ModalButtons's visibility */
  visibility: PropTypes.oneOf(['visible', 'disabled', 'hidden']),
  /** ModalButtons's id */
  id: PropTypes.string,
  /** additional external class name */
  className: PropTypes.string,
  /** Buttons */
  children: PropTypes.node.isRequired,
};

export default ModalButtons;
