// mw-style-react/build/Table/TableBody/TableBody.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _TableBody = require('./TableBody.style');

var _TableBody2 = _interopRequireDefault(_TableBody);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TableBody = (_dec = (0, _reactJss2.default)(_TableBody2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(TableBody, _PureComponent);

  function TableBody(props) {
    (0, _classCallCheck3.default)(this, TableBody);

    var _this = (0, _possibleConstructorReturn3.default)(this, (TableBody.__proto__ || (0, _getPrototypeOf2.default)(TableBody)).call(this, props));

    _this.state = {};
    return _this;
  }

  (0, _createClass3.default)(TableBody, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          classes = _props.classes,
          children = _props.children,
          columnSize = _props.columnSize;

      return _react2.default.createElement(
        'div',
        { className: (0, _clsx2.default)(classes.tableBody, className) },
        _react2.default.Children.map(children, function (child) {
          return _react2.default.cloneElement(child, {
            columnSize: columnSize
          });
        })
      );
    }
  }]);
  return TableBody;
}(_react.PureComponent)) || _class);


TableBody.defaultProps = {
  visibility: 'visible',
  columnSize: {}
};

TableBody.propTypes = {
  /** React-JSS injects a classes, which is a simple map of rule names and generated class names */
  classes: _propTypes2.default.object,
  /** additional class name */
  className: _propTypes2.default.string,
  /** table's Content */
  children: _propTypes2.default.node,
  /** table's columnSize Content */
  columnSize: _propTypes2.default.object
};

exports.default = TableBody;
