// mw-style-react/src/components/Banner/Banner.js
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import injectSheet from 'react-jss';
import cn from 'clsx';
import Utils from '../utils/utils';
import styles from './Banner.style';

@injectSheet(styles)
class Banner extends PureComponent {
  render() {
    const {classes, value, size, visibility, className, id, error} = this.props;

    return (
      <div
        id={id}
        className={cn(
          className,
          classes.banner,
          classes[size],
          {error: !!error},
          {hidden: visibility === 'hidden'},
        )}
        dangerouslySetInnerHTML={{__html: Utils.bbCodeToHtml(value)}}
      />
    );
  }
}

Banner.defaultProps = {
  visibility: 'visible',
  error: false,
  size: 'small',
};

Banner.propTypes = {
  classes: PropTypes.object,
  /** Banner's visibility */
  visibility: PropTypes.oneOf(['visible', 'hidden']),
  /** Does Banner's size */
  size: PropTypes.oneOf(['small', 'medium']),
  /** Does Banner's value (BBCode) */
  value: PropTypes.string.isRequired,
  /** Does Banner has an error */
  error: PropTypes.bool,
  /** Banner's id */
  id: PropTypes.string,
  /** additional external class name */
  className: PropTypes.string,
};

export default Banner;
