// mw-style-react/build/Label/Label.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _Label = require('./Label.style');

var _Label2 = _interopRequireDefault(_Label);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Label = (_dec = (0, _reactJss2.default)(_Label2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Label, _PureComponent);

  function Label() {
    (0, _classCallCheck3.default)(this, Label);
    return (0, _possibleConstructorReturn3.default)(this, (Label.__proto__ || (0, _getPrototypeOf2.default)(Label)).apply(this, arguments));
  }

  (0, _createClass3.default)(Label, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classes = _props.classes,
          fontSize = _props.fontSize,
          fontWeight = _props.fontWeight,
          textColor = _props.textColor,
          className = _props.className,
          value = _props.value,
          id = _props.id,
          error = _props.error,
          visibility = _props.visibility,
          onClick = _props.onClick,
          onMouseEnter = _props.onMouseEnter,
          onMouseLeave = _props.onMouseLeave;


      return _react2.default.createElement(
        'span',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.label, classes[fontSize], classes[fontWeight], { error: !!error }, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' }),
          style: { color: textColor },
          onClick: onClick,
          onMouseEnter: onMouseEnter,
          onMouseLeave: onMouseLeave
        },
        value
      );
    }
  }]);
  return Label;
}(_react.PureComponent)) || _class);


Label.defaultProps = {
  fontSize: 'medium',
  fontWeight: 'normal',
  visibility: 'visible',
  error: false,
  onClick: function onClick() {},
  onMouseEnter: function onMouseEnter() {},
  onMouseLeave: function onMouseLeave() {}
};

Label.propTypes = {
  classes: _propTypes2.default.object,
  /** label size */
  fontSize: _propTypes2.default.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge']),
  /** label weight */
  fontWeight: _propTypes2.default.oneOf(['normal', 'semibold']),
  /** label's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** Does label have an error */
  error: _propTypes2.default.bool,
  /** label custom color */
  textColor: _propTypes2.default.string,
  /** label's value */
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /** label's id */
  id: _propTypes2.default.string,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** label callback */
  onClick: _propTypes2.default.func,
  /** callback onMouseEnter */
  onMouseEnter: _propTypes2.default.func,
  /** callback onMouseLeave */
  onMouseLeave: _propTypes2.default.func
};

exports.default = Label;
