// mw-style-react/build/List/List.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _List = require('./List.style');

var _List2 = _interopRequireDefault(_List);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var List = (_dec = (0, _reactJss2.default)(_List2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(List, _PureComponent);

  function List() {
    (0, _classCallCheck3.default)(this, List);
    return (0, _possibleConstructorReturn3.default)(this, (List.__proto__ || (0, _getPrototypeOf2.default)(List)).apply(this, arguments));
  }

  (0, _createClass3.default)(List, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classes = _props.classes,
          id = _props.id,
          className = _props.className,
          visibility = _props.visibility,
          children = _props.children,
          activeItem = _props.activeItem,
          offset = _props.offset,
          onClick = _props.onClick;


      return _react2.default.createElement(
        'div',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.list, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' })
        },
        _react2.default.Children.map(children, function (child) {
          return _react2.default.cloneElement(child, {
            rootId: id,
            activeItem: activeItem,
            offset: offset,
            onClick: onClick
          });
        })
      );
    }
  }]);
  return List;
}(_react.PureComponent)) || _class);


List.defaultProps = {
  visibility: 'visible',
  offset: 12,
  onClick: function onClick() {}
};

List.propTypes = {
  classes: _propTypes2.default.object,
  /** List's options */
  children: _propTypes2.default.node.isRequired,
  /** menu's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']).isRequired,
  /** List's id */
  id: _propTypes2.default.string,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** active item */
  activeItem: _propTypes2.default.string,
  /** offset of list */
  offset: _propTypes2.default.number,
  /** onClick callback */
  onClick: _propTypes2.default.func
};

exports.default = List;
