// mw-style-react/.storybook/webpack.config.js
const path = require('path');
const marked = require('marked');
const webpack = require('webpack');

const renderer = new marked.Renderer();

renderer.image = function (href, title, text) {
  let size;
  if (title) {
    size = title.split('x');
    if (size[1]) {
      size = `width=${size[0]} height=${size[1]}`;
    } else {
      size = `width=${size[0]}`;
    }
  } else {
    size = '';
  }
  return ('<img src="' + href + '" alt="' + text + '" ' + size + '>');
};

module.exports = {
  module: {
    rules: [
      {
        test: /\.scss$/,
        loaders: ["style-loader", "css-loader", "sass-loader"],
        include: path.resolve(__dirname, '../')
      },
      {
        test: /\.(png|jpg|gif)$/,
        use: [
          {
            loader: 'file-loader',
            options: {}
          }
        ]
      },
      {
        test: /\.md$/,
        use: [
          {
            loader: "markdown-loader",
            options: {
              renderer,
              highlight: function (code) {
                return require('highlight.js').highlightAuto(code).value;
              }
            }
          }
        ]
      },
      {
        test: /\.svg$/,
        use: [{
          loader: 'svgr/webpack',
          options: {
            svgo: false,
          },
        }, 'file-loader'],
      }
    ]
  },
  plugins: [
    new webpack.DefinePlugin({
      __NODE_ENV__: JSON.stringify(process.env.NODE_ENV)
    })
  ]
};
