// mw-style-react/.storybook/customDecorators/infoDecorator.js
import React from "react";
import {withInfo} from '../addon-info/src/index';
import TableProps from '../components/TableProps/TableProps';

// addon-info
export default function infoDecorator(stories) {
  return stories.addDecorator((story, context) => {
    return withInfo({
      TableComponent: TableProps,
      styles: stylesheet => ({
        ...stylesheet,
        infoBody: {
          ...stylesheet.infoBody,
          border: 0,
          boxShadow: 0,
          padding: 0,
          fontFamily: 'Open Sans'
        },
        header: {
          ...stylesheet.header,
          body: {
            ...stylesheet.header.body,
          },
          h1: {
            ...stylesheet.header.h1,
            fontSize: 24
          },
          h2: {
            ...stylesheet.header.h2,
            fontSize: 18,
            marginTop: 10,
            display: 'none'
          }
        },
        source: {
          ...stylesheet.source,
          h1: {
            ...stylesheet.source.h1,
            fontSize: 18,
          }
        },
        propTableHead: {
          marginBottom: 10,
          fontSize: 18,
          display: 'none'
        }
      }),
      isWrapped: true
    })(story)(context)
  });
}
