// mw-style-react/src/components/Table/TableBody/TableBody.js
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import injectSheet from 'react-jss';
import cn from 'clsx';
import styles from './TableBody.style';

@injectSheet(styles)
class TableBody extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const {className, classes, children, columnSize} = this.props;
    return (
      <div className={cn(classes.tableBody, className)}>
        {React.Children.map(children, child =>
          React.cloneElement(child, {
            columnSize,
          }),
        )}
      </div>
    );
  }
}

TableBody.defaultProps = {
  visibility: 'visible',
  columnSize: {},
};

TableBody.propTypes = {
  /** React-JSS injects a classes, which is a simple map of rule names and generated class names */
  classes: PropTypes.object,
  /** additional class name */
  className: PropTypes.string,
  /** table's Content */
  children: PropTypes.node,
  /** table's columnSize Content */
  columnSize: PropTypes.object,
};

export default TableBody;
