// mw-style-react/build/Radio/RadioItem/RadioItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _RadioItem = require('./RadioItem.style');

var _RadioItem2 = _interopRequireDefault(_RadioItem);

var _Icon = require('../../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _Label = require('../../Label');

var _Label2 = _interopRequireDefault(_Label);

var _utils = require('../../utils/utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RadioItem = (_dec = (0, _reactJss2.default)(_RadioItem2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(RadioItem, _PureComponent);

  function RadioItem(props) {
    (0, _classCallCheck3.default)(this, RadioItem);

    var _this = (0, _possibleConstructorReturn3.default)(this, (RadioItem.__proto__ || (0, _getPrototypeOf2.default)(RadioItem)).call(this, props));

    _this.handleOnClick = _this.handleOnClick.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(RadioItem, [{
    key: 'handleOnClick',
    value: function handleOnClick() {
      var _props = this.props,
          value = _props.value,
          visibility = _props.visibility,
          onChange = _props.onChange;

      if (visibility === 'disabled') return;
      onChange(value);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          id = _props2.id,
          classes = _props2.classes,
          className = _props2.className,
          visibility = _props2.visibility,
          value = _props2.value,
          label = _props2.label,
          activeItem = _props2.activeItem,
          children = _props2.children;

      var uniqId = _utils2.default.random();
      var isChecked = activeItem === value;

      return _react2.default.createElement(
        'div',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.radioItem, { checked: isChecked }, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' })
        },
        _react2.default.createElement(
          'div',
          { className: classes.content },
          _react2.default.createElement('input', {
            className: classes.input,
            type: 'radio',
            value: isChecked,
            id: 'id-' + uniqId,
            onClick: this.handleOnClick
          }),
          _react2.default.createElement(_Icon2.default, {
            className: classes.icon,
            type: isChecked ? 'radio_on' : 'radio_off',
            size: 'medium',
            color: 'currentColor'
          }),
          _react2.default.createElement(
            'label',
            { className: classes.label, htmlFor: 'id-' + uniqId },
            _react2.default.createElement(_Label2.default, { value: label })
          )
        ),
        children && _react2.default.createElement(
          'div',
          { className: classes.children },
          children
        )
      );
    }
  }]);
  return RadioItem;
}(_react.PureComponent)) || _class);


RadioItem.defaultProps = {
  id: null,
  children: null,
  error: false,
  value: false,
  visibility: 'visible',
  onChange: function onChange() {}
};

RadioItem.propTypes = {
  classes: _propTypes2.default.object,
  /** RadioItem's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** RadioItem's label */
  label: _propTypes2.default.string.isRequired,
  /** RadioItem's child nodes */
  children: _propTypes2.default.node,
  /** RadioItem value */
  value: _propTypes2.default.string,
  /** active RadioItem */
  activeItem: _propTypes2.default.string,
  /** element id */
  id: _propTypes2.default.string,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** onChange callback */
  onChange: _propTypes2.default.func
};

exports.default = RadioItem;
