// mw-style-react/src/components/Modal/ModalContent/ModalContent.js
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import injectSheet from 'react-jss';
import cn from 'clsx';
import styles from './ModalContent.style';

@injectSheet(styles)
class ModalContent extends PureComponent {
  render() {
    const {classes, id, visibility, children, className} = this.props;

    return (
      <div
        id={id}
        className={cn(
          className,
          classes.contentWrapper,
          {disabled: visibility === 'disabled'},
          {hidden: visibility === 'hidden'},
        )}
      >
        <div className={classes.wrap}>{children}</div>
      </div>
    );
  }
}

ModalContent.defaultProps = {
  visibility: 'visible',
};

ModalContent.propTypes = {
  classes: PropTypes.object,
  /** ModalContent's visibility */
  visibility: PropTypes.oneOf(['visible', 'disabled', 'hidden']).isRequired,
  /** additional external class name */
  className: PropTypes.string,
  /** ModalContent's id */
  id: PropTypes.string,
  /** ModalContent's children */
  children: PropTypes.node.isRequired,
};

export default ModalContent;
