// mw-style-react/build/Notify/Notify.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _reactAddonsCssTransitionGroup = require('react-addons-css-transition-group');

var _reactAddonsCssTransitionGroup2 = _interopRequireDefault(_reactAddonsCssTransitionGroup);

var _Notify = require('./Notify.style');

var _Notify2 = _interopRequireDefault(_Notify);

var _NotifyItem = require('./NotifyItem');

var _NotifyItem2 = _interopRequireDefault(_NotifyItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Notify = (_dec = (0, _reactJss2.default)(_Notify2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Notify, _PureComponent);

  function Notify() {
    (0, _classCallCheck3.default)(this, Notify);
    return (0, _possibleConstructorReturn3.default)(this, (Notify.__proto__ || (0, _getPrototypeOf2.default)(Notify)).apply(this, arguments));
  }

  (0, _createClass3.default)(Notify, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          value = _props.value,
          onClose = _props.onClose,
          className = _props.className,
          classes = _props.classes;

      var reversedValue = value.slice(0).reverse();
      return _react2.default.createElement(
        'div',
        { className: (0, _clsx2.default)(classes.notify, className) },
        _react2.default.createElement(
          _reactAddonsCssTransitionGroup2.default,
          {
            transitionName: {
              enter: classes.animEnter,
              enterActive: classes.animEnterActive,
              leave: classes.animExit,
              leaveActive: classes.animExitActive
            },
            transitionEnterTimeout: 500,
            transitionLeaveTimeout: 500
          },
          reversedValue.map(function (item) {
            return _react2.default.createElement(_NotifyItem2.default, (0, _extends3.default)({ key: item.id, onClose: onClose }, item));
          })
        )
      );
    }
  }]);
  return Notify;
}(_react.PureComponent)) || _class);


Notify.defaultProps = {
  onClose: function onClose() {},
  value: []
};

Notify.propTypes = {
  classes: _propTypes2.default.object,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** Notify's value array with NotifyItem props object */
  value: _propTypes2.default.array,
  /** onClose callback */
  onClose: _propTypes2.default.func
};

exports.default = Notify;
