// mw-style-react/build/Badge/Badge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _Badge = require('./Badge.style');

var _Badge2 = _interopRequireDefault(_Badge);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Badge = (_dec = (0, _reactJss2.default)(_Badge2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Badge, _PureComponent);

  function Badge() {
    (0, _classCallCheck3.default)(this, Badge);
    return (0, _possibleConstructorReturn3.default)(this, (Badge.__proto__ || (0, _getPrototypeOf2.default)(Badge)).apply(this, arguments));
  }

  (0, _createClass3.default)(Badge, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classes = _props.classes,
          id = _props.id,
          className = _props.className,
          visibility = _props.visibility,
          value = _props.value,
          bgColor = _props.bgColor,
          size = _props.size,
          onClick = _props.onClick,
          borderColor = _props.borderColor;

      return _react2.default.createElement(
        'div',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.badge, classes[size], { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' }, 'badge'),
          onClick: onClick,
          style: {
            borderColor: borderColor,
            backgroundColor: bgColor
          }
        },
        value
      );
    }
  }]);
  return Badge;
}(_react.PureComponent)) || _class);


Badge.defaultProps = {
  visibility: 'visible',
  error: false,
  size: 'medium',
  onClick: function onClick() {}
};

Badge.propTypes = {
  classes: _propTypes2.default.object,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** Badge's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** Badge size */
  size: _propTypes2.default.oneOf(['small', 'medium', 'large', 'xlarge', 'xxlarge']),
  /** Badge's id */
  id: _propTypes2.default.string,
  /** Badge's value */
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /** Badge border color */
  borderColor: _propTypes2.default.string,
  /** Badge background color */
  bgColor: _propTypes2.default.string,
  /** callback onClick */
  onClick: _propTypes2.default.func
};

exports.default = Badge;
