// mw-style-react/build/Banner/Banner.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _utils = require('../utils/utils');

var _utils2 = _interopRequireDefault(_utils);

var _Banner = require('./Banner.style');

var _Banner2 = _interopRequireDefault(_Banner);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Banner = (_dec = (0, _reactJss2.default)(_Banner2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Banner, _PureComponent);

  function Banner() {
    (0, _classCallCheck3.default)(this, Banner);
    return (0, _possibleConstructorReturn3.default)(this, (Banner.__proto__ || (0, _getPrototypeOf2.default)(Banner)).apply(this, arguments));
  }

  (0, _createClass3.default)(Banner, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classes = _props.classes,
          value = _props.value,
          size = _props.size,
          visibility = _props.visibility,
          className = _props.className,
          id = _props.id,
          error = _props.error;


      return _react2.default.createElement('div', {
        id: id,
        className: (0, _clsx2.default)(className, classes.banner, classes[size], { error: !!error }, { hidden: visibility === 'hidden' }),
        dangerouslySetInnerHTML: { __html: _utils2.default.bbCodeToHtml(value) }
      });
    }
  }]);
  return Banner;
}(_react.PureComponent)) || _class);


Banner.defaultProps = {
  visibility: 'visible',
  error: false,
  size: 'small'
};

Banner.propTypes = {
  classes: _propTypes2.default.object,
  /** Banner's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'hidden']),
  /** Does Banner's size */
  size: _propTypes2.default.oneOf(['small', 'medium']),
  /** Does Banner's value (BBCode) */
  value: _propTypes2.default.string.isRequired,
  /** Does Banner has an error */
  error: _propTypes2.default.bool,
  /** Banner's id */
  id: _propTypes2.default.string,
  /** additional external class name */
  className: _propTypes2.default.string
};

exports.default = Banner;
