// mw-style-react/build/Table/TableHead/TableHead.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _TableHead = require('./TableHead.style');

var _TableHead2 = _interopRequireDefault(_TableHead);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TableHead = (_dec = (0, _reactJss2.default)(_TableHead2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(TableHead, _PureComponent);

  function TableHead(props) {
    (0, _classCallCheck3.default)(this, TableHead);

    var _this = (0, _possibleConstructorReturn3.default)(this, (TableHead.__proto__ || (0, _getPrototypeOf2.default)(TableHead)).call(this, props));

    _this.state = {
      columnSize: {}
    };
    return _this;
  }

  (0, _createClass3.default)(TableHead, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var columnSize = {};
      // Делаем один массив если часть tableCell добавляется динамически
      var newChildren = this.props.children.props.children.reduce(function (acc, e) {
        return acc.concat(e);
      }, []);

      newChildren.forEach(function (item, index) {
        columnSize[index] = item.props.width;
      });
      this.setState({
        columnSize: columnSize
      });
      this.props.getColumnSize(columnSize);
    }
  }, {
    key: 'render',
    value: function render() {
      var columnSize = this.state.columnSize;
      var _props = this.props,
          className = _props.className,
          classes = _props.classes,
          children = _props.children,
          fixed = _props.fixed;


      return _react2.default.createElement(
        'div',
        { className: (0, _clsx2.default)(classes.tableHead, className, { fixed: fixed }) },
        _react2.default.Children.map(children, function (child) {
          return _react2.default.cloneElement(child, {
            columnSize: columnSize
          });
        })
      );
    }
  }]);
  return TableHead;
}(_react.PureComponent)) || _class);


TableHead.defaultProps = {
  visibility: 'visible',
  getColumnSize: function getColumnSize() {},
  fixed: false
};

TableHead.propTypes = {
  /** React-JSS injects a classes, which is a simple map of rule names and generated class names */
  classes: _propTypes2.default.object,
  /** additional class name */
  className: _propTypes2.default.string,
  /** table's Content */
  children: _propTypes2.default.node,
  /** table's column size Content */
  getColumnSize: _propTypes2.default.func,
  /** table's header position*/
  fixed: _propTypes2.default.bool
};

exports.default = TableHead;
