// mw-style-react/build/Checkbox/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _Label = require('../Label');

var _Label2 = _interopRequireDefault(_Label);

var _utils = require('../utils/utils');

var _utils2 = _interopRequireDefault(_utils);

var _Checkbox = require('./Checkbox.style');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Checkbox = (_dec = (0, _reactJss2.default)(_Checkbox2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Checkbox, _PureComponent);

  function Checkbox(props) {
    (0, _classCallCheck3.default)(this, Checkbox);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Checkbox.__proto__ || (0, _getPrototypeOf2.default)(Checkbox)).call(this, props));

    _this.handleOnChange = _this.handleOnChange.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Checkbox, [{
    key: 'handleOnChange',
    value: function handleOnChange() {
      var _props = this.props,
          id = _props.id,
          value = _props.value,
          visibility = _props.visibility,
          required = _props.required,
          onChange = _props.onChange;

      if (visibility === 'disabled') return;
      var valueF = !_utils2.default.tryToBool(value);
      var error = required && !value;
      onChange({ id: id, value: valueF, error: error });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          id = _props2.id,
          classes = _props2.classes,
          className = _props2.className,
          error = _props2.error,
          visibility = _props2.visibility,
          value = _props2.value,
          children = _props2.children,
          helperText = _props2.helperText;

      var uniqId = _utils2.default.random();
      var isChecked = _utils2.default.tryToBool(value);

      return _react2.default.createElement(
        'div',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.checkbox, { checked: isChecked }, { error: !!error }, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' })
        },
        _react2.default.createElement(
          'div',
          { className: classes.content },
          _react2.default.createElement('input', {
            className: classes.input,
            type: 'checkbox',
            value: isChecked,
            id: 'id-' + uniqId,
            onChange: this.handleOnChange
          }),
          _react2.default.createElement(_Icon2.default, {
            className: classes.icon,
            type: isChecked ? 'checkbox_on' : 'checkbox_off',
            error: error,
            size: 'medium',
            color: 'currentColor'
          }),
          _react2.default.createElement(
            'label',
            { className: classes.label, htmlFor: 'id-' + uniqId },
            children
          )
        ),
        helperText && _react2.default.createElement(_Label2.default, {
          className: (0, _clsx2.default)(classes.helperText, { error: error }),
          value: helperText,
          fontSize: 'small',
          error: error
        })
      );
    }
  }]);
  return Checkbox;
}(_react.PureComponent)) || _class);


Checkbox.defaultProps = {
  visibility: 'visible',
  error: false,
  value: false,
  onChange: function onChange() {}
};

Checkbox.propTypes = {
  classes: _propTypes2.default.object,
  /** text field's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** checkbox child nodes */
  children: _propTypes2.default.node.isRequired,
  /** Does text field's has an error */
  error: _propTypes2.default.bool,
  /** checkbox helperText */
  helperText: _propTypes2.default.string,
  /** checkbox value */
  value: _propTypes2.default.bool,
  /** Is checkbox required? */
  required: _propTypes2.default.bool,
  /** element id */
  id: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /** additional external class name */
  className: _propTypes2.default.string,
  /** onChange callback */
  onChange: _propTypes2.default.func
};

exports.default = Checkbox;
