// mw-style-react/build/Radio/Radio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _Radio = require('./Radio.style');

var _Radio2 = _interopRequireDefault(_Radio);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Radio = (_dec = (0, _reactJss2.default)(_Radio2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(Radio, _PureComponent);

  function Radio(props) {
    (0, _classCallCheck3.default)(this, Radio);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Radio.__proto__ || (0, _getPrototypeOf2.default)(Radio)).call(this, props));

    _this.handleChange = _this.handleChange.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Radio, [{
    key: 'handleChange',
    value: function handleChange(value) {
      var _props = this.props,
          id = _props.id,
          visibility = _props.visibility,
          onChange = _props.onChange;

      if (visibility === 'disabled') return;
      var error = false;
      onChange({ id: id, value: value, error: error });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          id = _props2.id,
          classes = _props2.classes,
          className = _props2.className,
          value = _props2.value,
          children = _props2.children,
          visibility = _props2.visibility,
          align = _props2.align;

      return _react2.default.createElement(
        'div',
        {
          id: id,
          className: (0, _clsx2.default)(className, classes.radio, { hidden: visibility === 'hidden' }, { disabled: visibility === 'disabled' }, { horizontal: align === 'horizontal' })
        },
        _react2.default.Children.map(children, function (child) {
          return _react2.default.cloneElement(child, {
            className: classes.item,
            activeItem: value,
            onChange: _this2.handleChange
          });
        })
      );
    }
  }]);
  return Radio;
}(_react.PureComponent)) || _class);


Radio.defaultProps = {
  id: null,
  value: null,
  visibility: 'visible',
  align: 'vertical',
  error: false,
  onChange: function onChange() {}
};

Radio.propTypes = {
  classes: _propTypes2.default.object,
  /** Radio visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** Radio options */
  children: _propTypes2.default.node.isRequired,
  /** Radio align */
  align: _propTypes2.default.oneOf(['vertical', 'horizontal']),
  /** Radio value */
  value: _propTypes2.default.string,
  /** element id */
  id: _propTypes2.default.string,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** onChange callback */
  onChange: _propTypes2.default.func
};

exports.default = Radio;
