// mw-style-react/docs/pages/Brandbook/components/Buttons/Buttons.js
import React from 'react';
import PropTypes from 'prop-types';
import Utils from '../../../../../src/components/utils/utils';
import {Button} from '../../../../../src/components';
import './Buttons.scss';

const buttonsList = [
  [
    {
      size: 'small',
      label: 'Button',
      icon: 'search',
      iconAfter: 'arrow',
      visibility: 'disabled',
    },
    {
      size: 'small',
      label: 'Button',
      icon: 'search',
      visibility: 'disabled',
    },
    {
      size: 'small',
      label: 'Button',
      visibility: 'disabled',
    },
    {
      size: 'small',
      icon: 'search',
      visibility: 'disabled',
    },
    {
      size: 'small',
      icon: 'search',
      rounded: true,
      visibility: 'disabled',
    },
  ],
  [
    {
      size: 'small',
      label: 'Button',
      icon: 'search',
      iconAfter: 'arrow',
    },
    {
      size: 'small',
      label: 'Button',
      icon: 'search',
    },
    {
      size: 'small',
      label: 'Button',
    },
    {
      size: 'small',
      icon: 'search',
    },
    {
      size: 'small',
      icon: 'search',
      rounded: true,
    },
  ],
  [
    {
      size: 'smallplus',
      label: 'Button',
      icon: 'search',
      iconAfter: 'arrow',
    },
    {
      size: 'smallplus',
      label: 'Button',
      icon: 'search',
    },
    {
      size: 'smallplus',
      label: 'Button',
    },
    {
      size: 'smallplus',
      icon: 'search',
    },
    {
      size: 'smallplus',
      icon: 'search',
      rounded: true,
    },
  ],
  [
    {
      label: 'Button',
      icon: 'search',
      iconAfter: 'arrow',
    },
    {
      label: 'Button',
      icon: 'search',
    },
    {
      label: 'Button',
    },
    {
      icon: 'search',
    },
    {
      icon: 'search',
      rounded: true,
    },
  ],
  [
    {
      size: 'large',
      label: 'Button',
      icon: 'search',
      iconAfter: 'arrow',
    },
    {
      size: 'large',
      label: 'Button',
      icon: 'search',
    },
    {
      size: 'large',
      label: 'Button',
    },
    {
      size: 'large',
      icon: 'search',
    },
    {
      size: 'large',
      icon: 'search',
      rounded: true,
    },
  ],
];

const Buttons = ({type}) => (
  <div className="brandbook__buttons">
    {buttonsList
      .map(items => (
        <div className="brandbook__buttons__row" key={Utils.random()}>
          {items.map(item => (
            <div className="brandbook__buttons__row__item" key={Utils.random()}>
              <Button {...item} type={type} />
            </div>
          ))}
        </div>
      ))
      .reverse()}
  </div>
);

Buttons.defaultProps = {
  type: 'default',
};

Buttons.propTypes = {
  type: PropTypes.string,
};

export default Buttons;
