// mw-style-react/src/components/Autocomplete/AutocompleteItem/AutocompleteItem.style.js
export default (theme) => ({
  autocompleteItem: {
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
    padding: theme.autocomplete.item.paddingLarge,
    height: 56,
    lineHeight: theme.autocomplete.item.lineHeight,
    background: theme.autocomplete.item.background,
    transition: theme.transition(['background']),
    position: 'relative',
    '&.disabled': {
      opacity: theme.disabledOpacity,
      cursor: 'default'
    },
    '&.hidden': {
      display: 'none'
    },
    '&:hover': Object.assign(theme.autocomplete.item.hover, {}),
    '&.active': Object.assign(theme.autocomplete.item.hover, {}),
    '&.unActive': {
      '&:hover': {
        background: theme.autocomplete.item.background
      },
    },
    '& .nameContainer': {
      display: 'flex',
      flexGrow: 1,
      flexShrink: 1,
      flexBasis: 'auto',
      overflow: 'hidden',
      alignItems: 'center'
    },
    '& .label': {
      marginLeft: 12,
      color: theme.autocomplete.item.labelColor,
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
      overflow: 'hidden'
    },
    '&.medium': {
      height: 40,
      padding: theme.autocomplete.item.paddingMedium,
    }
  }
});
