// mw-style-react/build/themes/corezoid_light.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.palette2 = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Theme's main colors
 */
var palette = {
  primaryColor: '#3C91E6',
  secondaryColor: '#dfebf0',
  textColor: '#393e44',
  errorColor: '#e64646',
  secondaryErrorColor: '#ffe3dc',
  successColor: '#018a01',
  secondarySuccessColor: '#aff9af',
  disabledColor: '#dbdbdb',
  backgroundDisabledColor: '#EBEDEF',
  secondaryTextColor: '#ffffff',
  hoverMenuPrimaryColor: '#e6f4fc',
  backgroundActiveColor: '#e6f4fc',
  menuPrimaryColor: '#ffffff',
  tooltipColor: '#35404b',
  activePaginationColor: '#35404b',
  toggleSwitchNotActive: '#c2c5c9',
  modalBgColor: '#ffffff'
};

// Обновленная палитра
var palette2 = exports.palette2 = {
  primary: '#1973E8',
  primaryHover: '#1E6AD0',
  primary30: '#BAD5F8',
  primary10: '#F0F6FD',
  primary2: '#F9FBFE',
  white: '#FFFFFF',
  black: '#393F48',
  gray: '#7E8A9A',
  buttons: '#ACB3BE',
  disabled: '#DDE0E5',
  yellow: '#F4B400',
  green: '#129D59',
  purple: '#9037DB',
  red: '#DB4437',
  yellow10: '#FBE8B3',
  green10: '#B8DFCB',
  purple10: '#B8C0DF',
  red10: '#F1A7A7'
};

/**
 * Animation logic
 */
var easingArr = {
  easeInOut: 'cubic-bezier(.23,1,.32,1)',
  easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
  easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
  sharp: 'cubic-bezier(0.4, 0, 0.6, 1)'
};

/**
 * Transition wrapper
 */
function transition(props) {
  var duration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 450;
  var delay = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  var easing = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'easeInOut';

  var trArr = [];
  var easingVal = easingArr[easing];
  props.forEach(function (prop) {
    var tr = prop + ' ' + duration + 'ms ' + easingVal + ' ' + delay + 'ms';
    trArr.push(tr);
  });
  return trArr.join(', ');
}

/**
 * Transition wrapper for difficult animation
 */
function transitionArr(props) {
  var trArr = [];
  props.forEach(function (prop) {
    var field = prop[0] || 'all';
    var duration = prop[1] || 300;
    var delay = prop[2] || 0;
    var easingVal = easingArr[prop[3]] || easingArr.easeInOut;
    var tr = field + ' ' + duration + 'ms ' + easingVal + ' ' + delay + 'ms';
    trArr.push(tr);
  });
  return trArr.join(', ');
}

/**
 * boxShadow wrapper
 */
function boxShadow() {
  var level = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 2;

  var shadowLevels = {
    0: 'none',
    1: '0 2px 4px 0 rgba(57, 62, 68, 0.2)',
    2: '0 4px 7px 0 rgba(57, 62, 68, 0.2)',
    3: '0 8px 16px 0 rgba(57, 62, 68, 0.2)',
    4: '0 16px 24px 0 rgba(57, 62, 68, 0.2)',
    5: '0 24px 32px 0 rgba(57, 62, 68, 0.2)'
  };
  return shadowLevels[level];
}

/**
 * Base font
 */
var font = {
  family: {
    main: '"Open Sans", sans-serif;'
  },
  size: {
    small: 10,
    caption: 12,
    body: 14,
    subheader: 18,
    title: 24
  },
  weight: {
    regular: 400,
    semibold: 600
  }
};

/**
 * Main theme's settings
 */
var theme = {
  transition: transition,
  transitionArr: transitionArr,
  boxShadow: boxShadow,
  disabledOpacity: 0.5,
  disabledCursor: 'not-allowed',
  icon: {
    color: palette2.buttons,
    micro: 8,
    small: 12,
    medium: 16,
    large: 24,
    xlarge: 32
  },
  autocomplete: {
    borderColor: '#dbdbdb',
    borderWidthMedium: 1,
    boxShadow: boxShadow(),
    fontSizeMedium: font.size.body,
    searchBackgroundColor: palette2.white,
    placeholder: '#B5BCC2',
    rightTextColor: palette2.primary,
    item: {
      paddingLarge: '0px 16px 0 16px',
      paddingMedium: '0px 12px 0 12px',
      lineHeight: '40px',
      background: palette2.white,
      hover: {
        background: palette2.primary10
      },
      labelColor: '#242D37'
    }
  },
  avatar: {
    background: palette2.primary30,
    color: palette2.primary,
    size: {
      small: 24,
      medium: 32,
      large: 40,
      xlarge: 56,
      xxlarge: 80
    },
    fontSize: {
      small: font.size.small,
      medium: font.size.body,
      large: font.size.body,
      xlarge: font.size.subheader,
      xxlarge: font.size.subheader
    },
    fontWeight: {
      normal: font.weight.regular,
      semibold: font.weight.semibold
    }
  },
  button: {
    borderRadius: 4,
    border: 1,
    boxShadow: boxShadow(0),
    disabledTextColor: palette2.white,
    default: {
      normal: {
        backgroundColor: palette2.primary,
        color: palette2.white,
        borderColor: palette2.primary
      },
      hover: {
        backgroundColor: palette2.primaryHover,
        color: palette2.white,
        borderColor: palette2.primaryHover
      },
      disabled: {
        cursor: 'default',
        backgroundColor: palette2.disabled,
        borderColor: palette2.disabled,
        color: palette2.white
      }
    },
    secondary: {
      normal: {
        backgroundColor: palette2.white,
        color: palette2.primary,
        borderColor: palette2.primary
      },
      hover: {
        backgroundColor: palette2.primary10,
        color: palette2.primary,
        borderColor: palette2.primary
      },
      disabled: {
        cursor: 'default',
        backgroundColor: palette2.disabled,
        borderColor: palette2.disabled,
        color: palette2.white
      }
    },
    text: {
      normal: {
        backgroundColor: 'transparent',
        color: palette2.primary,
        borderColor: 'transparent'
      },
      hover: {
        backgroundColor: palette2.primary10,
        color: palette2.primary,
        borderColor: palette2.primary10
      },
      disabled: {
        cursor: 'default',
        backgroundColor: 'transparent',
        borderColor: 'transparent',
        color: palette2.disabled
      }
    }
  },
  chip: {
    background: palette2.primary2,
    borderColor: palette2.disabled,
    closeIconColor: palette2.buttons,
    maxLengthText: 120,
    small: {
      height: 24,
      rectangular: {
        borderRadius: 3
      }
    },
    medium: {
      height: 32,
      rectangular: {
        borderRadius: 4
      }
    },
    large: {
      height: 40,
      rectangular: {
        borderRadius: 6
      }
    }
  },
  label: {
    fontFamily: font.family.main,
    fontSize: {
      xsmall: font.size.small,
      small: font.size.caption,
      medium: font.size.body,
      large: font.size.subheader,
      xlarge: font.size.title
    },
    fontWeight: {
      normal: font.weight.regular,
      semibold: font.weight.semibold
    },
    errorColor: palette2.red
  },
  divider: {
    color: palette2.disabled
  },
  menu: {
    boxShadow: boxShadow(),
    bgColor: palette2.white,
    item: {
      bgColor: palette2.white,
      hover: {
        bgColor: palette2.primary10
      },
      active: {
        bgColor: palette2.primary30
      },
      label: {
        fontSize: font.size.body,
        color: palette2.black
      },
      children: {
        fontSize: font.size.caption,
        color: palette2.gray
      }
    }
  },
  modal: {
    small: 480,
    medium: 560,
    large: 640,
    borderRadius: 8,
    bgColor: palette.modalBgColor,
    boxShadow: boxShadow(),
    padding: 24,
    errorColor: palette.errorColor,
    spacingVertical: 24,
    spacingButton: 8,
    titleColor: palette2.black
  },
  notify: {
    width: 400,
    padding: 12,
    borderWidth: 1,
    borderRadius: 2,
    boxShadow: boxShadow(),
    textColor: palette2.black,
    bgColor: palette2.white,
    normalBorderColor: palette2.disabled,
    normalTitleBgColor: palette2.primary2,
    normalIconColor: palette2.buttons,
    successBorderColor: palette2.green,
    successTitleBgColor: palette2.green10,
    successIconColor: palette2.green,
    errorBorderColor: palette2.red,
    errorTitleBgColor: palette2.red10,
    errorIconColor: palette2.red
  },
  progressBar: {
    backgroundColor: palette2.disabled,
    color: palette2.primary,
    strokeWidth: 4
  },
  pagination: {
    labelColor: palette2.black,
    itemColor: palette2.black,
    itemHoverColor: palette2.primary,
    itemHoverBorderColor: palette2.primary,
    itemHoverBgColor: palette2.primary10,
    itemActiveColor: palette2.white,
    itemActiveBorderColor: palette2.primary,
    itemActiveBgColor: palette2.primary,
    itemBorderRadius: 2,
    fontSize: font.size.body,
    fontWeightSemibold: font.weight.semibold,
    iconColor: palette2.buttons,
    iconHoverColor: palette2.primary
  },
  tab: {
    borderBottomContainerColor: palette2.disabled,
    color: palette2.gray,
    activeColor: palette2.primary,
    errorColor: palette2.red,
    disabledColor: palette2.disabled,
    smallHeightUnderline: 1,
    mediumHeightUnderline: 2
  },
  textField: {
    borderColor: palette2.disabled,
    fontSizeLabel: font.size.body,
    colorLabel: palette2.gray,
    fontSizeHelper: font.size.caption,
    colorHelper: palette2.gray,
    fontSizeSmall: font.size.body,
    fontSizeMedium: font.size.body,
    placeholder: palette2.gray,
    closeIconColor: palette2.buttons,
    active: {
      fontSizeLabel: font.size.caption,
      colorLabel: palette2.primary,
      borderColor: palette2.primary,
      fontSizeHelper: font.size.caption,
      colorHelper: palette2.black,
      size: palette2.body,
      color: palette2.black,
      closeIconColor: palette2.buttons
    },
    selected: {
      fontSizeLabel: font.size.caption,
      colorLabel: palette2.gray,
      borderColor: palette2.buttons,
      fontSizeHelper: font.size.caption,
      colorHelper: palette2.black,
      size: palette2.body,
      color: palette2.black,
      closeIconColor: palette2.buttons
    },
    error: {
      fontSizeLabel: font.size.caption,
      colorLabel: palette2.gray,
      borderColor: palette2.red,
      fontSizeHelper: font.size.caption,
      colorHelper: palette2.red,
      size: palette2.body,
      color: palette2.black,
      closeIconColor: palette2.red
    },
    bordered: {
      borderRadiusSmall: 3,
      borderRadiusMedium: 4,
      borderRadiusLarge: 4
    }
  },
  toggleSwitch: {
    color: palette.toggleSwitchNotActive,
    activeColor: palette2.primary,
    borderColor: palette.toggleSwitchNotActive,
    borderActiveColor: palette2.primary,
    disabledColor: palette.disabledColor,
    errorColor: palette.errorColor,
    small: {
      width: 24,
      height: 12,
      border: 1
    },
    medium: {
      width: 28,
      height: 16,
      border: 2
    },
    large: {
      width: 34,
      height: 20,
      border: 2
    }
  },
  tooltip: {
    background: palette2.black,
    textColor: palette2.white,
    textSize: font.size.caption,
    borderRadius: 2
  },
  checkbox: {
    textColor: palette2.gray,
    textActiveColor: palette2.black,
    textDisabledColor: palette2.disabled,
    iconColor: palette2.buttons,
    iconActiveColor: palette2.primary,
    iconDisabledColor: palette2.disabled
  },
  list: {
    item: {
      active: {
        color: palette2.primary,
        background: palette.secondaryColor
      },
      hover: {
        color: palette2.primary
      }
    }
  },
  radio: {
    background: palette2.primary,
    color: palette.textColor,
    paddingHorizontal: 20,
    paddingLabel: 8,
    textColor: palette2.gray,
    textActiveColor: palette2.black,
    textDisabledColor: palette2.disabled,
    iconColor: palette2.buttons,
    iconActiveColor: palette2.primary,
    iconDisabledColor: palette2.disabled
  },
  sideBar: {
    item: {
      iconActiveColor: palette2.primary,
      textColor: palette2.gray,
      textActiveColor: palette2.black,
      bgColor: palette2.white,
      bgHoverColor: palette2.primary10,
      bgActiveColor: palette2.primary30,
      borderRadius: 4
    }
  },
  badge: {
    backgroundColor: palette2.primary,
    color: palette2.white,
    borderColor: palette2.white
  },
  upload: {
    backgroundColor: palette2.primary2,
    browseLinkColor: palette2.primary,
    iconUploadColor: palette2.black,
    iconCloseHoverFill: palette2.primary,
    borderRadius: 3,
    errorColor: palette2.red,
    borderColor: palette2.disabled,
    borderItemColor: '#cccccc'
  },
  code: {
    backgroundColor: palette2.primary10,
    borderRadius: 3,
    btnColor: palette2.primary,
    textColor: palette2.black,
    textSize: font.size.body
  },
  sendForm: {
    borderRadius: 3,
    borderColor: '#DEE3E8',
    iconFill: '#C0C9D1',
    active: {
      borderColor: palette2.primary
    }
  },
  calendar: {
    mainBgColor: palette2.white,
    periodColor: palette2.black,
    periodIconColor: palette2.black,
    periodIconHoverColor: palette2.primary,
    periodIconDisabledColor: palette2.disabled,
    dayColor: palette2.black,
    dayHoverColor: palette2.primary,
    dayHoverBorderColor: palette2.primary,
    dayHoverBgColor: palette2.primary10,
    dayActiveColor: palette2.primary,
    dayDisabledColor: palette2.disabled,
    dayOtherMonthColor: palette2.buttons,
    daySelectedColor: palette2.white,
    daySelectedBorderColor: palette2.primary,
    daySelectedBgColor: palette2.primary,
    boxShadow: boxShadow(),
    borderRadius: 3,
    timeBorderColor: palette2.buttons,
    timeBorderActiveColor: palette2.primary,
    timezoneIconColor: palette2.buttons
  },
  table: {
    rowActiveBgColor: palette2.primary,
    rowHoverBgColor: palette2.primary10,
    bgColor: palette2.white,
    cellFontSize: font.size.body,
    cellHeaderColor: palette2.gray,
    cellHeaderBgColor: palette2.white,
    cellColor: palette2.black,
    cellActiveColor: palette2.white,
    arrowColor: palette2.primary,
    arrowActiveColor: palette2.white
  },
  banner: {
    textColor: palette2.black,
    linkColor: palette2.primary,
    linkHoverColor: palette2.primaryHover,
    mainBgColor: palette2.primary10,
    errorBgColor: palette2.red10,
    mainLineBgColor: palette2.primary,
    errorLineBgColor: palette2.red,
    borderRadius: 2,
    fontSize: {
      small: font.size.caption,
      medium: font.size.body
    },
    fontWeight: {
      semibold: font.weight.semibold
    }
  }
};

exports.default = (0, _assign2.default)({}, theme, palette);
