// mw-style-react/src/components/Calendar/WeekDays/WeekDays.js
/**
 * Дни недели
 * */
import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import injectSheet from 'react-jss';
import cn from 'clsx';
import styles from './WeekDays.style';
import Label from '../../Label';

@injectSheet(styles)
class WeekDays extends PureComponent {
  render() {
    const {isoDays, classes, size, className} = this.props;
    const weekDayList = isoDays
      ? ['Mo', 'Tu', 'We', 'Tu', 'Fr', 'Sa', 'Su']
      : ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];
    return (
      <div className={cn(className, classes[size], classes.weekDayList)}>
        {weekDayList.map((weekDay, index) => (
          <Label key={index} fontSize={size} className={cn(classes.weekDay)} value={weekDay}/>
        ))}
      </div>
    );
  }
}

WeekDays.propTypes = {
  classes: PropTypes.object,
  /** additional external class name */
  className: PropTypes.string,
  /** WeekDays size */
  size: PropTypes.oneOf(['small', 'medium']).isRequired,
  /** WeekDays first day of the week is Monday by ISO-8601 */
  isoDays: PropTypes.bool.isRequired,
};

export default WeekDays;
