// mw-style-react/build/Tab/TabItem/TabItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _class;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _clsx = require('clsx');

var _clsx2 = _interopRequireDefault(_clsx);

var _TabItem = require('./TabItem.style');

var _TabItem2 = _interopRequireDefault(_TabItem);

var _index = require('../../Label/index');

var _index2 = _interopRequireDefault(_index);

var _index3 = require('../../Icon/index');

var _index4 = _interopRequireDefault(_index3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TabItem = (_dec = (0, _reactJss2.default)(_TabItem2.default), _dec(_class = function (_PureComponent) {
  (0, _inherits3.default)(TabItem, _PureComponent);

  function TabItem(props) {
    (0, _classCallCheck3.default)(this, TabItem);

    var _this = (0, _possibleConstructorReturn3.default)(this, (TabItem.__proto__ || (0, _getPrototypeOf2.default)(TabItem)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(TabItem, [{
    key: 'handleClick',
    value: function handleClick() {
      var _props = this.props,
          value = _props.value,
          visibility = _props.visibility,
          onChange = _props.onChange;

      if (visibility === 'disabled') return;
      onChange(value);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          id = _props2.id,
          classes = _props2.classes,
          value = _props2.value,
          visibility = _props2.visibility,
          leftIcon = _props2.leftIcon,
          rightIcon = _props2.rightIcon,
          leftIconSize = _props2.leftIconSize,
          rightIconSize = _props2.rightIconSize,
          label = _props2.label,
          size = _props2.size,
          type = _props2.type,
          error = _props2.error,
          activeItem = _props2.activeItem,
          itemRef = _props2.itemRef,
          className = _props2.className,
          tabItemIndent = _props2.tabItemIndent;


      return _react2.default.createElement(
        'div',
        {
          ref: itemRef,
          id: id,
          className: (0, _clsx2.default)(className, classes.tab, classes[size], classes[type], classes['indent' + tabItemIndent], { error: !!error }, { disabled: visibility === 'disabled' }, { hidden: visibility === 'hidden' }, { active: activeItem === value }, 'tabItem'),
          onClick: this.handleClick
        },
        leftIcon && _react2.default.createElement(_index4.default, {
          type: leftIcon,
          className: (0, _clsx2.default)(classes.icon, classes.leftIcon, 'leftIcon'),
          size: leftIconSize
        }),
        label && _react2.default.createElement(_index2.default, {
          className: (0, _clsx2.default)(classes.label),
          value: label,
          fontSize: size === 'small' ? 'small' : 'medium'
        }),
        rightIcon && _react2.default.createElement(_index4.default, {
          type: rightIcon,
          className: (0, _clsx2.default)(classes.icon, classes.rightIcon, 'rightIcon'),
          size: rightIconSize
        })
      );
    }
  }]);
  return TabItem;
}(_react.PureComponent)) || _class);


TabItem.defaultProps = {
  id: null,
  error: false,
  visibility: 'visible',
  type: 'fixed',
  size: 'medium',
  tabItemIndent: 'medium',
  leftIconSize: 'small',
  rightIconSize: 'small',
  onChange: function onChange() {}
};

TabItem.propTypes = {
  classes: _propTypes2.default.object,
  /** TabItem's visibility */
  visibility: _propTypes2.default.oneOf(['visible', 'disabled', 'hidden']),
  /** TabItem's id */
  id: _propTypes2.default.string,
  /** TabItem's value */
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]).isRequired,
  /** active item */
  activeItem: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /** TabItem's callback */
  onChange: _propTypes2.default.func,
  /** TabItem's leftIcon */
  leftIcon: _propTypes2.default.string,
  /** TabItem's leftIcon */
  leftIconSize: _propTypes2.default.oneOf(['small', 'micro']),
  /** TabItem's rightIcon */
  rightIcon: _propTypes2.default.string,
  /** TabItem's leftIcon */
  rightIconSize: _propTypes2.default.oneOf(['small', 'micro']),
  /** TabItem's label */
  label: _propTypes2.default.string,
  /** TabItem's size */
  size: _propTypes2.default.oneOf(['small', 'medium']),
  /** TabItem's type */
  type: _propTypes2.default.string,
  /** Error Tab id */
  error: _propTypes2.default.bool,
  /** TabItem's reference */
  itemRef: _propTypes2.default.func,
  /** additional external class name */
  className: _propTypes2.default.string,
  /** TabItem indent */
  tabItemIndent: _propTypes2.default.oneOf(['small', 'medium', 'large'])
};

exports.default = TabItem;
